### ISSUE TITLE
Allow transform to recursively search for BAMs in a directory and convert them all to ADAM

### ISSUE THREAD

**Issue by ryan-williams** [bigdatagenomics/adam#741]
simply swapping `s/\.bam$/.adam/` on the file names.

I feel like I/we end up not converting stuff to ADAM because I have a mental block about doing it file by file

**Comment by fnothaft** [comment]
Would you intend this to write out "one" parquet file (a directory of parquet files that was all of the samples), or to convert each .bam into an individual .adam parquet file? To rephrase, if you started with 100 BAM files, would you end with 100 ADAM files or 1 ADAM file?

**Comment by ryan-williams** [comment]
ah, I was imagining 100 ADAM files, all dropped in place next to their respective BAMs

**Comment by laserson** [comment]
This seems like a departure from the typical Hadoopy semantics.  If I give a Hadoop engine an input path in HDFS, I typically expect one output path to go along with it.

**Comment by massie** [comment]
I think what Ryan is looking for here (correct me if I'm wrong) is something like a recursive `sync` option. A Hadoopy option here might be to specify a `bam_directory` and `adam_directory` and have a job that checks which BAM files in the `bam_directory` haven't been converted and convert them to ADAM using the `adam_directory` as the root of the output. Agree with @laserson here that mixing the file types together together would complicate things. We would, for example, need to filter out all the BAM files on input (which is possible but not ideal).

**Comment by ryan-williams** [comment]
I'm open to peoples' preferences about what the directory structure should be, though a default of dropping an `.adam` next to any `.bam` feels really sane and useful to me.

The most common pattern at HammerLab is to do this, similarly to how other things in bioinformatics (e.g. `.bai`s) are often associated with their "parent"/"source" files by living next to them in a directory and sharing a basic `basename` structure, with a different extension.

So, most of the ADAM `transform` commands I run look like this:

```
bin/adam-submit transform /path/to/foo.{b,ad}am
```

I'm "this close" to writing my own crappy script to do this recursively in an HDFS directory, to all the directories or files ending in `.bam` that it finds, but I would much prefer ADAM to do it for me, not least of all because I imagine it would be able to benefit greatly from parallelizing N such `transform`ations in one Spark application instead of launching one Spark application per `.bam`.

It is common for us to have a folder on HDFS that has many sub-folders (corresponding to e.g. individual patients), each of which has multiple/many BAMs. One thing keeping us from lifting all of our analyses into ADAM-space, right off of the bat, is that we'd have to run 100s of `adam transform`s, and that feels annoying to script and wait on, today.

**Comment by heuermh** [comment]
At risk of veering off-topic, here is a stub of such a script written for [Nextflow](http://nextflow.io), which has some nice properties in terms of process isolation and abstraction of execution environment.

https://github.com/heuermh/bdg-nextflow/blob/master/transform.nf

As I learn more about ADAM usage patterns I hope to tighten the interface between Nextflow and BDG applications.  At this point ADAM would be treated as any other command line application, which as you say is not the most efficient way to execute such a workflow.

**Comment by ryan-williams** [comment]
**tl;dr:** what if I add back `bam2adam` and give it recursive capability, as well as the MD-tagging capability from #790?

I'm in a degrading orbit around this issue and have an implementation question.

As I alluded to in #790, one thing that I think would make this simpler is to make each `foo2adam` and `adam2foo` command able to operate recursively; it's pretty straightforward to see how that would work: traverse directory structure looking for specific extensions, convert them where found.

Having `transform` attempt to do this seems a little more complicated: should any `.bam`s/`.vcf`s/`.fa`s/`.bed`s found be converted to ADAM? Should any ADAM files that are found be converted back to `.bam`/`.vcf`/`.fa`/`.bed`/etc.? How do we indicate that we are in "convert things to ADAM" mode vs. "convert things from ADAM"? More cmdline flags to `transform`?

How do we feel about eventually having explicit `foo2adam` and `adam2foo` commands for all supported types, and having them be able to operate recursively, instead of trying to have `transform` do it? My current feeling is that this would be best. `transform` can stick around being as smart as possible about what you're asking it to do, but the format-specific commands will be able to include functionality that would be potentially more clumsy to tell `transform` about.

Today we have:
* `adam2fastq`, no `fastq2adam`
* `fasta2adam` no `adam2fasta`
* `features2adam`, no `adam2features` (related: #710)
* `vcf2adam` and `adam2vcf`
* neither `bam2adam` nor `adam2bam`
  * an apparently local-only `bam2adam` was removed in #674 
  * #689 started by adding `adam2bam` but then I suggested it be rolled into `transform` :)

I mostly want this recursive capability for `bam2adam`; recursively adding MD-tags per #790 would also be great. I think I'm ok with us having a couple of ways to get similar functionality from the cmdline, where each command does some things that others don't and in all cases they are thin wrappers over underlying functionality that can be exposed a few different ways.

So I guess my proposal, then, is to add a proper `bam2adam` with recursive-conversion capabilities. Thoughts?

**Comment by fnothaft** [comment]
That sounds reasonable to me. Should we get this merged before doing the larger refactor with `bam2adam`? Further comments below:

* Perhaps we could add `fastq2adam` as part of #771? 
* I've opened #793 for `adam2fasta`
* I think the recursive conversion sounds cool, but how do you actually plan to do that conversion? Would it make sense to use something like the [Kite SDK](http://kitesdk.org/docs/current/) and to add dataset ID fields to the schema? If not, would there be a good way to do this _without_ doing it serially?
* Perhaps it'd make sense to merge this as part of `transform`, and then refactor this to be part of `bam2adam`?

**Comment by ryan-williams** [comment]
Well I haven't actually started writing any code for this yet, so no rush on merging the `transform` version :)

However, what I'm trying to say is that I'm not even sure how the `transform` version should work.

```
adam-submit transform /path/to/dir             #1
adam-submit transform -r /path/to/dir          #2
adam-submit transform -bam2adam /path/to/dir   #3
```

`transform` requires `<INPUT>` and `<OUTPUT>` paths today, which doesn't really map to what I'm trying to do.

I'd just as soon add a `bam2adam` with a `-r` flag.

> how do you actually plan to do that conversion? 

Something like:
```
for {
  path <- traverseDirectoryRecursivelyLookingForExtensions(inputDir, ".bam", ".sam")
  rdd = sc.loadBam(path)
} { 
  rdd.adamParquetSave(path.dropRight(4) + ".adam") 
}
```

Does that make sense?

**Comment by fnothaft** [comment]
> Well I haven't actually started writing any code for this yet, so no rush on merging the transform version 

Oy, sorry for the confusion! I meant that we should get #790 in as part of `transform` instead of waiting for `bam2adam`.

> Something like: <snip>

Ah, so then it'd be a serial, file-by-file approach?

I just realized that I was being a bit silly about something here. I was all like "wouldn't running conversions serially be slow?" but then it hit me that Spark would parallelize the independent serial runs (since they don't have between stage dependencies), so it is probably safe to ignore most of my blabbering higher up on the ticket. Sorry, have been coffee deprived today...

**Comment by ryan-williams** [comment]
> I meant that we should get #790 in as part of transform instead of waiting for bam2adam.

Yes, I agree, let's do that. Moreover I'd be OK leaving it accessible from `transform` and `bam2adam` in the future.

re: serial-ness: yes, I should have called this aspect of my thinking out more explicitly. I'm assuming there will be performance wins as Spark parallelizes all the `transform`s automatically. I'm pretty excited to try it out, feels like what some of these tools are really made for. I have 100s of BAMs laying around I'd like to just convert in one fell swoop.

**Comment by fnothaft** [comment]
+1 to the above; I think we can sort out any remaining questions as they come up.

**Comment by ryan-williams** [comment]
I sat down to work on this today and have now talked myself out of the idea that any of this can be parallelized easily in Spark.

AFAIK all the `RDD.save*` methods initiate blocking actions; is there a way to tell Spark to write multiple files at the same time?

Some ink has been spilled [on the mailing list](http://mail-archives.us.apache.org/mod_mbox/spark-dev/201507.mbox/%3CCA+XRB6GdXPDbQpyvO+DnbWN55RbMxwJx=xfT3ZbixRq1PwrhRw@mail.gmail.com%3E) and [on JIRA](https://issues.apache.org/jira/browse/SPARK-3533) about something that I think is equivalent: writing a keyed-RDD to multiple (per-key) files, and it seems pretty solidly "won't-fix"ed, so I think we're stuck here, unless I'm misunderstanding.

**Comment by ryan-williams** [comment]
I should add that I actually think it might still be worth adding recursive capability to `transform` or analogous format-specific commands, and I'm looking into doing that.

Finally, I think I can do this inside `transform` after all; if the `-recursive` flag is provided, then `<INPUT>` will be expected to be of the form `<dir>/**.<extension>` and `<OUTPUT>` will be expected to be just an extension, and any input paths with the specified input extension will be converted to have the output extension.
